src/aggregated_range_proof/dealer.rs

```diff
@@ -0,0 +1,237 @@
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use curve25519_dalek::traits::Identity;
+use generators::GeneratorsView;
+use inner_product_proof;
+use proof_transcript::ProofTranscript;
+use std::clone::Clone;
+use util;
+
+use super::messages::*;
+
+/// Dealer is an entry-point API for setting up a dealer
+pub struct Dealer {}
+
+impl Dealer {
+    /// Creates a new dealer with the given parties and a number of bits
+    pub fn new(
+        n: usize,
+        m: usize,
+        transcript: &mut ProofTranscript,
+    ) -> Result<DealerAwaitingValues, &'static str> {
+        if !n.is_power_of_two() {
+            return Err("n is not a power of two");
+        }
+        if !m.is_power_of_two() {
+            return Err("m is not a power of two");
+        }
+        transcript.commit_u64(n as u64);
+        transcript.commit_u64(m as u64);
+        Ok(DealerAwaitingValues { n, m })
+    }
+}
+
+/// When the dealer is initialized, it only knows the size of the set.
+#[derive(Debug)]
+pub struct DealerAwaitingValues {
+    n: usize,
+    m: usize,
+}
+
+impl DealerAwaitingValues {
+    /// Combines commitments and computes challenge variables.
+    pub fn receive_value_commitments(
+        self,
+        value_commitments: &Vec<ValueCommitment>,
+        transcript: &mut ProofTranscript,
+    ) -> Result<(DealerAwaitingPoly, ValueChallenge), (DealerAwaitingValues, &'static str)> {
```

Comment:
The error type of this `Result` is a `(DealerAwaitingValues, &'static str)`. I guess this is so that if there's an error, the code orchestrating the dealer state can retry and keep going with the protocol.

Is it a good idea to try to resume the protocol from an intermediate dealer state, or to fail without the possibility of resumption (i.e., return just the error part `&'static str`, and consume the dealer state so that the protocol has to start over)?

I think failing without resumption would be simpler, and I'm not sure whether it would make sense to retry here, since the error is caused by not having enough value commitments (which seems like something the orchestrating code should be ensuring when it collects them from whereever).